graylog2-server/src/main/java/org/graylog/scheduler/DBJobTriggerService.java

```diff
@@ -169,12 +169,23 @@ public List<JobTriggerDto> getAllForJob(String jobDefinitionId) {
             throw new IllegalArgumentException("jobDefinitionId cannot be null or empty");
         }
 
-        final Query query = DBQuery.is(FIELD_JOB_DEFINITION_ID, jobDefinitionId);
-        try (final DBCursor<JobTriggerDto> cursor = db.find(query)) {
+        try (final DBCursor<JobTriggerDto> cursor = db.find(forJobDefinitionQuery(jobDefinitionId))) {
             return ImmutableList.copyOf(cursor.iterator());
         }
     }
 
+    public JobTriggerDto upsertForJobDefinition(JobTriggerDto dto, String jobDefinitionId) {
+        if (isNullOrEmpty(jobDefinitionId)) {
+            throw new IllegalArgumentException("jobDefinitionId cannot be null or empty");
+        }
+        final BasicDBObject none = new BasicDBObject();
+        return db.findAndModify(forJobDefinitionQuery(jobDefinitionId), none, none, false, dto, true, true);
```

Comment:
I think we can't perform a `findAndModify` with a replacement `JobTriggerDto` object here. It will clobber the trigger lock. We have to update individual fields and leave the lock intact.

Reply:
Maybe something along the lines of https://github.com/Graylog2/graylog2-server/blob/81e34a5b17d91c9f8428a83bce7612eedd3e2a26/graylog2-server/src/main/java/org/graylog/scheduler/DBCustomJobDefinitionService.java#L56 could work.

Reply:
@bernd @thll Thinking through this more and looking at other examples in the code... Maybe a good way forward would be to execute a standard update call for the trigger (updating only the specific needed fields), and get the write result with number of entities affected (should be one or zero). If that is zero, then insert a new trigger in a new call. The main benefit is that it performs the query for an existing trigger and the update in a single call, reducing the likelihood of updating stale data be updating the whole trigger DTO object with a stale instance (which was being done initially, and is being done in the current `EventDefinitionHandler`). So, this seems like it would be an improvement. What do you think? If there are any concerns with that, there is also the option of deleting the trigger, and inserting a new one (seem to be what is done by `ReportSchedulingHandler`).

I don't see much benefit in trying to combine the insert/update of the trigger into an upsert in a single step, since that seems to only help with the case that two users update an existing scanner to enable sync at the same time, which is probably quite unlikely to happen. But, definitely open to feedback that.

Reply:
Actually, the `update` parameter of `findOneAndUpdate` might support setting both a `$set` for specific fields for updating an existing matching record, and the `$setOnInsert` for the new record insertion case. If so, that would handle the update of just a few fields on existing records, or inserting a whole new document, which would be ideal. I'll check if that works this afternoon.

https://www.mongodb.com/docs/manual/reference/operator/update/#fields

Reply:
> Actually, the `update` parameter of `findOneAndUpdate` might support setting both a `$set` for specific fields for updating an existing matching record, and the `$setOnInsert` for the new record insertion case

Actually, it looks like that is not supported in Mongo. Please see comment here:  https://github.com/Graylog2/graylog-plugin-enterprise/pull/7153#issuecomment-2093855932

Reply:
As part of the changes described [here](https://github.com/Graylog2/graylog-plugin-enterprise/pull/7153#issuecomment-2096935062), I rolled-back the upsert methods and added a new method to find or create a job definition by config field/value.